### PR TITLE
support hpa strategies

### PR DESCRIPTION
**Pull request by tpiperatgod**
fix issue: #530

### CODE REVIEW

File: src/main/java/io/functionmesh/compute/util/FunctionsUtil.java [hunk 1]
```diff
@@ -429,6 +434,44 @@ public static V1alpha1Function createV1alpha1FunctionFromFunctionConfig(String k
         }
         specPod.setEnv(env);
 
+        // Handle HPA Configurations
+        if (customRuntimeOptions.getHpaConfig() != null && customRuntimeOptions.getMaxReplicas() > 1) {
+            HPAConfig hpaConfig = customRuntimeOptions.getHpaConfig();
```
Comment (freeznet) [inline src/main/java/io/functionmesh/compute/util/FunctionsUtil.java:439]: could you please introduce a new function that converts `HPAConfig` to `BuiltinAutoscaler` and `AutoScalingBehavior` so we can reduce the complexity of `createV1alpha1FunctionFromFunctionConfig`?

File: src/main/java/io/functionmesh/compute/models/CustomRuntimeOptions.java [hunk 2]
```diff
@@ -39,11 +39,8 @@ public class CustomRuntimeOptions {
...
     private String runnerImage;
 
-    @TrustedModeOnly
     private String logLevel;
-
-    @TrustedModeOnly
     private String logRotationPolicy;
-
     private String runnerImageTag;
+    private HPAConfig hpaConfig;
```
Comment (freeznet) [inline src/main/java/io/functionmesh/compute/models/CustomRuntimeOptions.java:45]: this definition should keep similarity as the https://github.com/streamnative/function-mesh-worker-service/pull/542 did, could you please work with @jiangpengcheng to make sure we have similar naming for both VPA and HPA? thanks.

**Review by freeznet — APPROVED** [review]
please check

File: src/main/java/io/functionmesh/compute/util/FunctionsUtil.java (hunk 1 above)
Comment (jiangpengcheng) [reply src/main/java/io/functionmesh/compute/util/FunctionsUtil.java:439]: need to do the "reverse" change in `createFunctionConfigFromV1alpha1Function`

**Comment by tpiperatgod** [comment]
samples:

```shell
bin/pulsar-admin functions create --namespace default --tenant public --name py-fn-demo \
--auto-ack true \
--classname exclamation_function.ExclamationFunction \
--custom-runtime-options '{"maxReplicas":5,"runnerImage":"streamnative/pulsar-functions-python-runner","runnerImageTag":"2.9.3.18","hpaSpec":{"builtinCPURule":"AverageUtilizationCPUPercent80","builtinMemoryRule":"AverageUtilizationMemoryPercent80"}}' \
--py /pulsar/examples/python-examples/exclamation_function.py \
--inputs "persistent://public/default/input-python-topic" \
--output "persistent://public/default/output-python-topic" \
--log-topic "persistent://public/default/py-function-logs"
```

```shell
bin/pulsar-admin functions create --namespace default --tenant public --name py-fn-demo \
--auto-ack true \
--classname exclamation_function.ExclamationFunction \
--custom-runtime-options '{"maxReplicas":5,"runnerImage":"streamnative/pulsar-functions-python-runner","runnerImageTag":"2.9.3.18","hpaSpec":{"autoScalingMetrics":[{"type":"Resource","resource":{"name":"cpu","target":{"type":"Utilization","averageUtilization":80}}}],"autoScalingBehavior":{"scaleDown":{"stabilizationWindowSeconds":120,"policies":[{"type":"Percent","value":100,"periodSeconds":15}]},"scaleUp":{"stabilizationWindowSeconds":120,"policies":[{"type":"Percent","value":50,"periodSeconds":15},{"type":"Pods","value":2,"periodSeconds":15}],"selectPolicy":"Max"}}}}' \
--py /pulsar/examples/python-examples/exclamation_function.py \
--inputs "persistent://public/default/input-python-topic" \
--output "persistent://public/default/output-python-topic" \
--log-topic "persistent://public/default/py-function-logs"
```

File: src/main/java/io/functionmesh/compute/models/CustomRuntimeOptions.java [hunk 3]
```diff
@@ -39,11 +39,8 @@ public class CustomRuntimeOptions {
     @TrustedModeOnly
     private String runnerImage;
 
-    @TrustedModeOnly
```
Comment (jiangpengcheng) [inline src/main/java/io/functionmesh/compute/models/CustomRuntimeOptions.java:42]: any reason to remove this?

Comment (tpiperatgod) [reply src/main/java/io/functionmesh/compute/models/CustomRuntimeOptions.java:42]: `logLevel` should be a customizable parameter

**Comment by jiangpengcheng** [comment]
in the `createFunction(Sink|Source)ConfigFromV1alpha1Function(Sink|Source)`, we should convert k8s spec to functionConfig(sinkConfig|sourceConfig)

**Review by freeznet — APPROVED** [review]

**Comment by freeznet** [comment]
@tpiperatgod could you please fix the conflict, thanks.

**Comment by tpiperatgod** [comment]
> in the `createFunction(Sink|Source)ConfigFromV1alpha1Function(Sink|Source)`, we should convert k8s spec to functionConfig(sinkConfig|sourceConfig)

```shell
# bin/pulsar-admin functions get --name py-fn-demo --namespace default --tenant public
{
  "tenant": "public",
  "namespace": "default",
  "name": "py-fn-demo",
  "className": "exclamation_function.ExclamationFunction",
  "inputs": [
    "persistent://public/default/input-python-topic"
  ],
  "inputSpecs": {
    "persistent://public/default/input-python-topic": {
      "isRegexPattern": false,
      "poolMessages": false
    }
  },
  "output": "persistent://public/default/output-python-topic",
  "producerConfig": {},
  "logTopic": "persistent://public/default/py-function-logs",
  "processingGuarantees": "ATLEAST_ONCE",
  "forwardSourceMessageProperty": true,
  "userConfig": {},
  "runtime": "PYTHON",
  "autoAck": true,
  "parallelism": 1,
  "resources": {
    "cpu": 1.0,
    "ram": 1181116006,
    "disk": 10737418240
  },
  "jar": "function://public/default/py-fn-demo",
  "py": "download/pulsar_functions/exclamation_function.py",
  "cleanupSubscription": true,
  "customRuntimeOptions": "{\"clusterName\":\"standalone\",\"inputTypeClassName\":\"[B\",\"outputTypeClassName\":\"[B\",\"maxReplicas\":5,\"managed\":true,\"hpaSpec\":{\"builtinCPURule\":\"AverageUtilizationCPUPercent80\",\"builtinMemoryRule\":\"AverageUtilizationMemoryPercent80\"}}",
  "subscriptionPosition": "Latest"
}
```

**Review by freeznet — APPROVED** [review]

**Pull request merged**